### ISSUE TITLE
Support for dashes in type property names

### ISSUE THREAD

**Issue by flixcheck** [rhys-vdw/ts-auto-guard#128]
Currently, it is not possible to use a dash in a type property name, e.g.

```typescript
export type Foo = {
  "foo-1": number,
  "bar-2": string
}
```

It **is** possible to use dashes in **interface** property names, though.

A (commented out) test for this is already in place in [PR 127](https://github.com/rhys-vdw/ts-auto-guard/pull/127).

**Comment by rhys-vdw** [comment]
The system we have in place (checking for quotes in the property name) works some of the time, which suggests this is an upstream inconsistency in ts-morph.

Otherwise we could consider restoring this behaviour that @0xorial introduced to fix this issue: https://github.com/rhys-vdw/ts-auto-guard/pull/118/commits/7a4433bb0526262588153fbcb0a1a57edc001fc2

Or we could just always use indexer syntax for accessing properties.

**Comment by flixcheck** [comment]
What's the downside of always using indexer syntax? The files are auto-generated and shouldn't be changed manually anyway. The logical result is the same, but with less bugs / possible problems.

**Comment by chrisczach** [comment]
The only thing I can think of is that most linter configurations are going to want to have it changed to dot notation on the generated code but that's really non-issue since it's going to need to be linted anyways.

**Comment by flixcheck** [comment]
Yes you are right, _most_ linter configurations! It is an opinion of the linter configurator (to be honest, we also have the quotes option on "when needed"). In our case, we're firing prettier to prettify the files after they have been generated by ts-auto-guard.

**Comment by rhys-vdw** [comment]
Yeah, I mean I try to have it generate sensible looking code, but you can just set your linter to ignore `guard.ts`, which makes sense since you shouldn't be modifying the generated files.

**Comment by ArdorDeosis** [comment]
I just wanted to open a similar issue, then found this one. The problem goes far beyond hyphens. From a quick look at it it looks like the checks for single or double quotes are not working, only the checks for spaces and parseInt. (https://github.com/rhys-vdw/ts-auto-guard/blob/261bc691b2c293675e4eff5c05c05ed7ae8a19d7/src/index.ts#L703)

I created a small reporoduction project: https://github.com/ArdorDeosis/ts-auto-guard-bug-recreation

This type
```
export type SomeType = {
	'key': string,
	'a-hyphen': string,
	'a whitespace': string,
	'#hashtag': string,
	'key.with.dots': string,
	'0startingWithNumber': string,
	'"quotes inside my key"': string,
	"'quotes inside my key'": string
}
```
results in this guard
```
export function isSomeType(obj: any, _argumentName?: string): obj is SomeType {
    return (
        (obj !== null &&
            typeof obj === "object" ||
            typeof obj === "function") &&
        typeof obj.key === "string" &&
        typeof obj.a - hyphen === "string" &&
        typeof obj["a whitespace"] === "string" &&
        typeof obj.#hashtag === "string" &&
        typeof obj.key.with.dots === "string" &&
        typeof obj["0startingWithNumber"] === "string" &&
        typeof obj["quotes inside my key"] === "string" &&
        typeof obj["'quotes inside my key'"] === "string" &&
        typeof obj["more quotes"] === "string" &&
        typeof obj["'more quotes"] === "string" &&
        typeof obj.new
line === "string"
)
}
```

I think maybe a regex check would be better instead of the check for quotes.

I've never contributed to a public repo before, if I'd want to tackle this issue on the weekend, how would I go about that? Just cloning, changing and then opening a PR? Or do I fork the project?

**Comment by ArdorDeosis** [comment]
just for shits and giggles I (successfully) tried to inject code via a property key.

```typescript
export type InjectedCode = {
	"key": string,
	"key===\"string\"&&\n//_execute_injected_code:\n(()=>{\nconsole.log('injection_successful');\nreturn('string');\n})()": string,
}
```
results in 
```typescript
export function isInjectedCode(obj: any, _argumentName?: string): obj is InjectedCode {
    return (
        (obj !== null &&
            typeof obj === "object" ||
            typeof obj === "function") &&
        typeof obj.key === "string" &&
        typeof obj.key === "string" &&
        //_execute_injected_code:
        (() => {
            console.log('injection_successful');
            return ('string');
        })() === "string"
    )
}
```

**Comment by ArdorDeosis** [comment]
I made a PR: https://github.com/rhys-vdw/ts-auto-guard/pull/184
There are still property names that are not covered by this, including property names with double quotes `"`, backslashes `\` or newline characters `\n` in them, but many more are supported with this change, including dashes.

Unfortunately I wasn't able to link the PR to this issue after I opened it and did not think to do so while I created it, not sure if that would have been possible.

**Comment by rhys-vdw** [comment]
> Unfortunately I wasn't able to link the PR to this issue after I opened it and did not think to do so while I created it, not sure if that would have been possible.

FYI You can edit the message and add `Closes #128`. I'll do it for you now.
